### PR TITLE
OpenTelemetry Metrics for Middlewares

### PR DESCRIPTION
**Pull request by ajatprabha**
Add OTel implementations of middlewares for metrics support.

### CODE REVIEW

File: otelcourier/publish_test.go [hunk 1]
```diff
@@ -95,6 +84,38 @@ func TestPublishTraceSpan(t *testing.T) {
 		assert.Equal(t, publishErrMessage, span.Status().Description)
 	})
 
+	t.Run("Metrics", func(t *testing.T) {
+		vsn := courier.Version()
+		buf := bytes.NewBufferString(fmt.Sprintf(`# HELP courier_publish_attempts_total Number of publish attempts
+# TYPE courier_publish_attempts_total counter
+courier_publish_attempts_total{mqtt_qos="1",mqtt_retained="false",mqtt_topic="test-topic",otel_scope_name="github.com/gojek/courier-go/otelcourier",otel_scope_version="semver:%s",service_name="test-service"} 1
```
Comment (dhruvjain99) [inline otelcourier/publish_test.go:91]: mqtt_topic label can have non-deterministic values which can break the prometheus limit.

Comment (dhruvjain99) [reply otelcourier/publish_test.go:91]: Same goes for subscribe and unsubscribe events.

**Review by dhruvjain99 — CHANGES_REQUESTED** [review]
Can we also instrument number of connected clients?

File: otelcourier/publish_test.go (hunk 1 above)
Comment (ajatprabha) [reply otelcourier/publish_test.go:91]: Addressed with a new `Option`.

```go
// TopicAttributeTransformer helps transform topic before making an attribute for it.
// It is used in metric recording only. Traces use the original topic.
type TopicAttributeTransformer func(context.Context, string) string
```

**Review by dhruvjain99 — APPROVED** [review]

**Pull request merged**